geode-redis/src/distributedTest/java/org/apache/geode/redis/OutOfMemoryDUnitTest.java

```diff
@@ -0,0 +1,175 @@
+    serverProperties = new Properties();
+    locatorProperties.setProperty(MAX_WAIT_TIME_RECONNECT, "15000");
+
+    locator = clusterStartUp.startLocatorVM(0, locatorProperties);
+    server1 = clusterStartUp.startRedisVM(1, serverProperties, locator.getPort());
+    server2 = clusterStartUp.startRedisVM(2, serverProperties, locator.getPort());
+
+    server1.getVM().invoke(() -> {
+      RedisClusterStartupRule.getCache().getResourceManager().setCriticalHeapPercentage(5.0F);
+    });
+    server2.getVM().invoke(() -> {
+      RedisClusterStartupRule.getCache().getResourceManager().setCriticalHeapPercentage(5.0F);
+    });
+
+    redisServerPort1 = clusterStartUp.getRedisPort(1);
+    redisServerPort2 = clusterStartUp.getRedisPort(2);
+
+    jedis1 = new Jedis(LOCAL_HOST, redisServerPort1, JEDIS_TIMEOUT);
+    jedis2 = new Jedis(LOCAL_HOST, redisServerPort2, JEDIS_TIMEOUT);
+
+    char[] largeCharData = new char[VALUE_SIZE];
+    Arrays.fill(largeCharData, 'a');
+    valueString = new String(largeCharData);
+
+  }
+
+  @Before
+  public void testSetup() {
+    jedis1.flushAll();
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    jedis1.disconnect();
+    jedis2.disconnect();
+
+    server1.stop();
+    server2.stop();
+  }
+
+  @Test
+  public void shouldReturnOOMError_forWriteOperations_whenThresholdReached() {
+    IgnoredException.addIgnoredException(expectedEx);
+    IgnoredException.addIgnoredException("LowMemoryException");
+
+    fillMemory(jedis1, MAX_ITERATION_COUNT, false);
+
+    assertThatThrownBy(() -> jedis2.set("oneMoreKey", "value")).hasMessageContaining("OOM");
+  }
+
+  @Test
+  public void shouldAllowDeleteOperations_afterThresholdReached() {
+    IgnoredException.addIgnoredException(expectedEx);
+    IgnoredException.addIgnoredException("LowMemoryException");
+
+    fillMemory(jedis1, MAX_ITERATION_COUNT, false);
+
+    assertThatNoException().isThrownBy(() -> jedis2.del(FILLER_KEY + 1));
+  }
+
+  @Test
+  public void shouldAllowExpiration_afterThresholdReached() {
+    IgnoredException.addIgnoredException(expectedEx);
+    IgnoredException.addIgnoredException("LowMemoryException");
+
+    fillMemory(jedis1, MAX_ITERATION_COUNT, true);
+
+    GeodeAwaitility.await().until(() -> jedis2.ttl(FILLER_KEY + 1) == -2);
+  }
+
+  // TODO: test that write operations become allowed after memory has dropped
+  // below critical levels. Difficult to do right now because of vagaries of the
+  // Java garbage collector.
+
+  private int fillMemory(Jedis jedis, int maxIterations, boolean withExpiration) {
+    int i = 0;
+    while (i < maxIterations) {
+      try {
+        if (withExpiration) {
+          jedis.setex(FILLER_KEY + i, KEY_TTL_SECONDS, valueString);
+        } else {
+          jedis.set(FILLER_KEY + i, valueString);
+        }
+      } catch (JedisException je) {
+        assertThat(je).hasMessageContaining("OOM command not allowed");
+        break;
+      }
+      i++;
+    }
```

Comment:
It looks like the `shouldReturnOOMError_forWriteOperations_whenThresholdReached` is failing in CI because the OOM error never gets thrown.  I'm wondering if we're not filling the memory entirely.  Maybe you could do a while(hasMemoryAvailable){} loop (`hasMemoryAvailable` being a boolean variable that's initially set to `true`) then set `hasMemoryAvailable` to `false` when the `JedisException` is caught to ensure the memory is full beforehand.

Reply:
It's more complicated than that. The latest iteration does a GC before the 'fill' operation, and adds progressively smaller keys to cram the memory as full as possible. It still fails every so often, apparently because we don't have fine enough control over other usages of the JVM memory.

Reply:
Might a possible solution be to spin off a separate thread that continuously adds keys to maintain memory pressure, so that even if something does free up some memory after we first hit the OOM error in `fillMemory()`, we're able to get back to a low memory situation?

Reply:
I added such a thread - maybe not entirely elegantly yet - and altered fillMemory() in a couple ways. First, it does multiple forceGC() calls, to help ensure that we're filling the memory with our Redis data. Secondly, it loops multiple times with progressively smaller chunks of data, to help ensure every nook and cranny is filled.

Haven't seen it fail YET, but I have to admit I'm still worried about flakiness.